### ISSUE TITLE
Suggested Features/Functions

### ISSUE THREAD

**Issue by timlem** [Stelagmite/learning-management-system#1]
Here's where to start...

What core features are you looking for with an LMS system inside of Rock?

**Comment by timlem** [comment]
Here are some notes I already had as I was working on it myself for my church:
1. Signup individually or as a group
2. Have categories
3. Have tags
4. Image
5. Description
6. Viewable by people in a group/dataview
7. Have questions
8. Restrict access to next one after completing first one
9. Ability to have time spans (the module is open from specific date range)
10.Have video(s)
11. Have resources (pdf's, etc)
12. Use Steps to measure how many people have signed up for, started, completed

**Comment by asdf1nit** [comment]
I agree with all that @timlem said and have built most of that out for us. I would like to add a couple things that I feel are important as well. Core has a lot of things that can be Frankenstein'd together but there could be a few tweaks that can be made with minimal impact to how things already work and yet be useful for managing this as well.

- Have a system that is manageable by non rock admin users in your org with little to no involvement outside of initial setup.
- Check-in type system that keeps attendance, as well as the ability to have Groups\Classes\Content that doesn't require it.
- Storage. How is this stuff stored? Lets face it everyone has a different strategy, what does that look like with this?
- Front end blocks and user blocks.

**Comment by markewampler** [comment]
I am not trying to derail the conversation here completely but to suggest a possible different direction on this overall...

Instead of making this a `Core` thing it became a plugin and it was just hooks into a currently existing LMS.

I would think overall that would be a better bang for you buck in regards to continuing updates and improvements and less management for the Core team to worry about.  The LMS would continue to innovate in their space to keep their product relevant and it would not require any funds directed at Rock to get a better product.

We are working on an integration with Pathwright LMS.  It would require pathwright to come to the table and open up their API some more, but they would bite more at a potential audience increase that could be available for minimal effort on their part.

**Comment by jsw-austin** [comment]
> I am not trying to derail the conversation here completely but to suggest a possible different direction on this overall...
> 
> Instead of making this a `Core` thing it became a plugin and it was just hooks into a currently existing LMS.
> 
> I would think overall that would be a better bang for you buck in regards to continuing updates and improvements and less management for the Core team to worry about. The LMS would continue to innovate in their space to keep their product relevant and it would not require any funds directed at Rock to get a better product.
> 
> We are working on an integration with Pathwright LMS. It would require pathwright to come to the table and open up their API some more, but they would bite more at a potential audience increase that could be available for minimal effort on their part.

To me this idea maximizes extensibility (aligned with the values of Rock)  and has the added benefit (possibly) of taking some of the burden of management off the core team.

**Comment by charlottedean** [comment]
I agree with Mark but also like what Daniel has suggested as well.  My organization has also been in talks with Leadr, which is more staff specific but in the same ballpark with everything we are all talking about.

**Comment by cabal95** [comment]
At this point, it's probably still too early to decide on integration vs plugin vs core. The early goal should be to get a list of what bare minimum requirements each church has. That will really be the driving factor of the decision. The bare minimum should be kept to just that. Basically the list of what features it must have for you to even consider it.

For example, with a decision on whether to choose Rock or a different system would Rock not having integration with background checks be a deal breaker? Meaning, if background checks were something you needed and even though you could run them manually via an external site and upload the completed background check into Rock manually - would that be a deal breaker? Would you throw away the entire "basket of eggs" (Rock) just because it lacked that one feature?

When talking to my own staff, I usually try to get the list of requirements in 3 chunks:

1. The absolute bare minimum (what I described above) features it MUST have to even consider the project.
2. The "we really WANT these but they probably aren't deal breakers" items.
3. Finally the "would be NICE" things.

> It would be NICE if Rock had a learning management system and we WANT it to integrate with our background check service but it MUST let us do electronic children's checkin.

Once those are in place, it becomes much easier to make decisions like should this be a core feature in Rock, an integration with another system(s), or a plugin that gets built.

**Comment by markewampler** [comment]
<ol>
<li>
<p>our MVP is as follows:</p>
<p>&nbsp;</p>
<ol>
<li>A group member of a specifically chosen GroupType(s) needs to be able to launch a group without staff input
<ol>
<li>When launched allow launcher to select who will be participating</li>
<li>Allow for non-group members to join as well
<ol>
<li>In case they want a friend to join that is not a part of the group</li>
</ol>
</li>
<li>Allow for a &ldquo;Moderator/Leader&rdquo; to be chosen who is responsible for the groups progress
<ol>
<li>Chosen among group members or ley leadership over group or staff leadership over group
<ol>
<li>Leadership are stored in attributes about the group</li>
</ol>
</li>
</ol>
</li>
</ol>
</li>
<li>Allow staff to have an ongoing &ldquo;registration&rdquo; for group placement into a particular course/class
<ol>
<li>Random members register for a potluck type group</li>
<li>Staff member curates group and launches when a sufficient number of members have registered and a &ldquo;Moderator/Leader&rdquo; is available to serve</li>
</ol>
</li>
<li>Allow Staff/Leaders to launch a curated group
<ol>
<li>Present a &ldquo;registration&rdquo; url that anyone who has it can join the group</li>
</ol>
</li>
<li>Allows Videos as a task</li>
<li>Allows Audio as a task</li>
<li>Allows Images as a task</li>
<li>Allows PDF&rsquo;s as a task</li>
<li>Has Assigned Tasks
<ol>
<li>Allow Groups of Tasks to be assigned as due at the same time</li>
<li>Required Tasks</li>
<li>Optional Tasks
<ol>
<li>Optional do not count towards &ldquo;Completing the course&rdquo;</li>
</ol>
</li>
<li>Tasks can be time gated
<ol>
<li>Task 1 is available(unlocked)&nbsp;starting on Jan 1<sup>st</sup> then Task 2 is available(unlocked)&nbsp;on Jan 8<sup>th</sup> and so on</li>
<li>Always backwards available to complete old tasks</li>
</ol>
</li>
</ol>
</li>
<li>Has a forum for the group to discuss<br />
<ol>
<li>posts and replies</li>
<li>Unique to that group and only that group can see it
<ol>
<li>Allows for more vulnerability in sharing</li>
</ol>
</li>
</ol>
</li>
<li>Automated Emails at set times
<ol>
<li>Sends to group members reminding of upcoming tasks due</li>
<li>Sends to group members reminding of tasks becoming available</li>
<li>Welcome Emails</li>
<li>Closing Emails</li>
</ol>
</li>
<li>Allow for Moderator/Leader to send emails if needed outside of automated emails
<ol>
<li>To whole group</li>
<li>To individuals in group</li>
<li>Record all emails on person record (CYA in case of abuse)</li>
</ol>
</li>
<li>&ldquo;Quizzes&rdquo; or questions about material</li>
<li>Ability to create Reports on group data
<ol>
<li>Individual Completion percentage</li>
<li>Group completion percentage</li>
<li>Furthest Task completed if incomplete</li>
</ol>
</li>
<li>Allows historical access to material if I have completed a class/course previously
<ol>
<li>forever or until Jesus comes back</li>
</ol>
</li>
<li>Automatically&nbsp;updates record showing that &ldquo;Person A&rdquo; has completed &ldquo;Course B&rdquo; on X Date</li>
<li>Have different courses and security on courses that are available based on previous completion of other courses</li>
</ol>
</li>
</ol>

**Comment by marcierempel** [comment]
I'll have my team review the necessary requirements and post them here. 

We currently are operating with a learning management system that we had built about 2 years ago using the features in Rock. It's only used for video training at the moment, using group requirements, person attributes, dataviews, content channel and custom template on a content channel view block. 

I can share my documentation here if anyone wants to see it, but I'm sure it's a little clunkier than what you're thinking of doing. It doesn't allow for collaboration, questions, or group interaction and that's what we would like to see as we move more towards online groups and classes.

**Comment by timlem** [comment]
@markewampler thank you for that level of detail!

I agree with Daniel that we should start with what the MVP is and then decide if it should be core or a plugin or how to proceed from there. So everyone take a look at features listed so far and then if there is something you see a need for that is not yet listed feel free to add as a comment on this discussion. 

I am going to create another discussion for specific use cases. How would you plan on using the LMS. I will post mine and if there are any other use cases add them as comments

**Comment by marcierempel** [comment]
Our MVP would be the same as Mark has listed with the addition of the following (more for the Learning Designers & Content Authors):

These are not deal breakers but would be nice to have:

- Learning Maps with objective, sequencing and navigation options

- Common modules that can be referenced in multiple courses.

**Comment by TheArkChurch** [comment]
So I'll reference a group\class\training\course all as "course" for this outline. 

Our MVP requirements are:

The ability to create a course that has the following attributes\properties
  - Signup individually or as a group(family).
  - Have categories.
  - Has a Status\State that is configurable. Started, Pending, Completed
  - Course Image, Description, Start & Stop Dates, Public Flag, Teacher\Leader
  - Extendable with other attributes
  - Ability to make sign up\login required and\or completely anonymous with no login 
  - Ability to have individual sessions with start and stop time, as well as the ability to manually order sessions.
  - The ability to mark\complete a single step in a step program (single status\single linkage)
  - Workflow triggers
  - Ability to archive members
  - Check-in system if configured, this would require a login for the course

Sessions should have the following attributes\properties
  - Videos( multiple )
  - Documents( multiple )
  - Ability to make Inactive
  - Descriptions
  - Prerequisite sessions
  - Ability to mark a session completed
  - Ability to mark\complete a single step in a program (single status\single linkage)
  - Extendable with other attributes.
  - Workflow triggers

Other Requirements
  - Same login as rock.(people have a hard enough time remembering passwords if we make them have another login to something else... well that's a very hard sell, and more than likely a no)
  - Ability to be administered by non-rock admin after initial setup with little to no involvement for course creation\setup
  - Reminders for leaders\members that are configurable 
  - Leader communication to all course members

Future abilities of this should include
  - Have questions both at the group\class\training\course level and session level. (bonus for grading answers)
  - Ability for courses to be nested to allow making a course\track
  - Prerequisite courses for nested courses or any course
  - Ability to make a session change the State\Status of the course\class
  - Ability to make a Video required, and determine completion by either the amount of the video watched and/or questions answered
  - Assigned tasks
  - Required videos
  - Reporting
  - Workflow actions
  - The ability to mark\complete multiple steps(multiple linkages) in a step program with status(conditional)
